pkgs/native_assets_builder/doc/protocol/v1.md

```diff
@@ -0,0 +1,231 @@
+Build Hook Specification
+========================
+
+Version 1: As of 2024-01-24 implemented behind `--enable-experiment=native-assets`
+
+### Concepts
+
+#### Asset
+
+An asset is a file together with metadata that tells the Dart and Flutter SDK how to bundle such asset in an application bundle.
+
+Currently, all such assets are dynamic libraries.
+
+An asset must have an `assetId`. Dart code that uses an asset, references the asset using the `assetId`.
+
+An asset must have a target os and target architecture. When accessing the asset at runtime, the asset for the current os and architecture is accessed.
+
+An asset must have a link mode. Currently, the only supported link mode is dynamic. (Static linking of native code will be added later.)
+
+An asset must specify a path type for dynamic linking. 
+
+* `absolute` paths are absolute paths in the output dir (see below).
+* `system` paths are expected to resolve on the target machine PATH. In this case the asset is not a file but only metadata.
+* `process` "paths" have no path, symbols are resolved in the current process. In this case the asset is not a file but only metadata.
```

Comment:
Yeah so code assets are already much more involved than "a file". Code assets can be metadata only. And some variants of that metadata do not require any path.

path_type is probably not the right name.

Maybe it should be dynamic_library_type. But the executable and process one don't even have a dynamic library.

So then it should be dynamic_linking_type. This also sounds weird, because the linking always is the same in a way.

Maybe it should be dynamic_resolution_type. That gets closer. But the problem here is that we're also abstracting over the resolution partially. Dart and Flutter repackage dynamic libraries. So in this protocol for absolute paths, it's the absolute path at the point of this protocol. But for system paths, process, and executable the metadata gets passed in unmodified.

So then it should maybe be something like:

```yaml
- assetId: ...
  dynamic_resolution:
    type: absolute
  path: ... # Not nested in dynamic_resolution, it's the path for the protocol.
- assetId: ...
  dynamic_resolution:
    type: system
    path: ... # Nested in dynamic_resolution, it's the metadata to be passed into linking.
- assetId: ...
  dynamic_resolution:
    type: process
- assetId: ...
  dynamic_resolution:
    type: executable
```

wdyt?